common/robotPositioner.h

```diff
@@ -0,0 +1,186 @@
+
+
+class RobotPositioner {
+
+private:
+    // Robot variables
+    millimeter_t m_pos_X;                                           // robot's x position
+    millimeter_t m_pos_Y;                                           // robot's y position
+    millimeter_t m_pos_Z;                                           // robot's z position
+    millimeter_t m_goalPositionX;                                   // goal position x
+    millimeter_t m_goalPositionY;                                   // goal position y
+    millimeter_t m_goalPositionZ;                                   // goal position z
+    millimeter_t m_distanceFromGoal;                                // Euclidean distance from goal coordinate
+    degree_t m_heading;                                             // heading (angle) of the robot
+    degree_t m_goalAngle;                                           // angle to turn after finding the correct location
+    degree_t m_bearingFinal;                                        // heading converted to range <-180, 180> from <0, 360>
+    degree_t m_bearingFromGoal;                                     // bearing (angle) from goal coordinate
+
+    millimeter_t m_robot_r;                                         // robot's wheel radius
+    millimeter_t m_robot_D;                                         // robots' axis length
+    millimeter_t m_stopping_distance;                               // if the robot's distance from goal < stopping dist, robot stops
+    degree_t m_allowed_heading_error;                               // the amount of error allowed in the final heading
+    double m_k1;                                                    // curveness of the path to the goal
+    double m_k2;                                                    // speed of turning on the curves
+    double m_alpha;                                                 // causes more sharply peaked curves 
+    double m_beta;                                                  // causes to drop velocity if 'k'(curveness) increases
+    meters_per_second_t m_max_velocity;                             // max velocity
+
+    // robot superclass
+    BoBRobotics::Robots::Tank &m_bot;   
+  
+
+    
+    void driveRobotWithVelocities(const meters_per_second_t v,
+                                  const degrees_per_second_t w,
+                                  const millimeter_t robot_r,
+                                  const millimeter_t robot_d) {
+        // v = r * (Vl+Vr)/2
+        // w = r * (Vr-Vl)/D
+        double a =  (robot_r/2).value(); //convert to cm
+        double b =  (robot_r/2).value();
+        double c = ( robot_r / robot_d).value();
+        double d = (-(robot_r / robot_d).value());
+
+        double e=(v.value()); // converting to millimeters/second 
+        double f=w.value();
+        double Vl,Vr;
+
+        // determinant
+        double det = (a*d - b*c);
+    
+        Vl = ((e*d - b*f)/det);
+        Vr = ((a*f - e*c)/det);
```

Comment:
Finally, all these intermediate variables can be marked as ``const`` to make it clear that you don't change them once you calculate them

Reply:
I think ``double det = (a*d - b*c);`` is equivalent to ``double det = -2 * a * c;``, right?

I would also be inclined not to bother converting all these values to doubles -- you could stick with the unit-typed variables, write ``auto det = [whatever]``, then the units will cancel out in the next calculation for ``Vl`` and ``Vr`` and I *think* you can then just treat them as doubles. But this is up to you.

Reply:
I did initially write a comment suggesting this as well but I'm semi-sure that Vl and Vr don't come out as reasonable units i.e. it's trying to do something like distance/time + angle/time

Reply:
Ahh yes, my bad. You're totally right